airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/state/SyncManager.kt

```diff
@@ -11,30 +11,51 @@ import io.airbyte.cdk.command.DestinationCatalog
 import io.airbyte.cdk.command.DestinationStream
 import io.airbyte.cdk.message.Batch
 import io.airbyte.cdk.message.BatchEnvelope
+import io.airbyte.cdk.write.StreamLoader
 import io.github.oshai.kotlinlogging.KotlinLogging
 import io.micronaut.context.annotation.Factory
+import io.micronaut.context.annotation.Secondary
 import jakarta.inject.Singleton
 import java.util.concurrent.ConcurrentHashMap
 import java.util.concurrent.atomic.AtomicBoolean
 import java.util.concurrent.atomic.AtomicLong
+import kotlinx.coroutines.CompletableDeferred
 import kotlinx.coroutines.channels.Channel
 
 /** Manages the state of all streams in the destination. */
-interface StreamsManager {
+interface SyncManager {
     /** Get the manager for the given stream. Throws an exception if the stream is not found. */
-    fun getManager(stream: DestinationStream.Descriptor): StreamManager
+    fun getStreamManager(stream: DestinationStream.Descriptor): StreamManager
+
+    fun registerStartedStreamLoader(streamLoader: StreamLoader)
+    suspend fun getOrAwaitStreamLoader(stream: DestinationStream.Descriptor): StreamLoader
 
     /** Suspend until all streams are closed. */
     suspend fun awaitAllStreamsClosed()
 }
 
-class DefaultStreamsManager(
+class DefaultSyncManager(
     private val streamManagers: ConcurrentHashMap<DestinationStream.Descriptor, StreamManager>
-) : StreamsManager {
-    override fun getManager(stream: DestinationStream.Descriptor): StreamManager {
+) : SyncManager {
+    private val streamLoaders =
+        ConcurrentHashMap<DestinationStream.Descriptor, CompletableDeferred<StreamLoader>>()
+
+    override fun getStreamManager(stream: DestinationStream.Descriptor): StreamManager {
         return streamManagers[stream] ?: throw IllegalArgumentException("Stream not found: $stream")
     }
 
+    override fun registerStartedStreamLoader(streamLoader: StreamLoader) {
+        streamLoaders
+            .getOrDefault(streamLoader.stream.descriptor, CompletableDeferred())
```

Comment:
Nice catch, thank you!

Tests aren't failing because the blocking behavior isn't in the task launcher any more, so it's not tested.

I added a test case to SyncManagerTest